---
description: "Prefer `this` return type over explicit class name when returning `this`."
title: "returnThisTypes"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="returnThisTypes" />

[Method chaining](https://en.wikipedia.org/wiki/Method_chaining) is a pattern in object-oriented programming where class methods return the instance of the class being called.
TypeScript provides a special polymorphic `this` type to facilitate method chaining.
When a base class method returns `this`, calling that method on any sub-classes will be known to return the specific sub-class, rather than the general base class.

Class methods that explicitly declare a return type of the class name instead of this make it harder for extending classes to call that method: the returned object will be typed as the base class, not the derived class.

This rule reports on class methods that declare their return type as that specific class name rather rather than `this`.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
class Builder {
	setValue(): Builder {
		return this;
	}
}
```

```ts
class Builder {
	setValue = (): Builder => this;
}
```

```ts
class Builder {
	get self(): Builder {
		return this;
	}
}
```

```ts
class Animal<T> {
	eat(): Animal<T> {
		return this;
	}
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
class Builder {
	setValue(): this {
		return this;
	}
}
```

```ts
class Builder {
	setValue = (): this => this;
}
```

```ts
class Builder {
	get self(): this {
		return this;
	}
}
```

```ts
class Builder {
	clone(): Builder {
		return new Builder();
	}
}
```

```ts
class Builder {
	setValue() {
		return this;
	}
}
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you intentionally want to restrict method chaining to the exact class type and prevent subclasses from inheriting the fluent interface, this rule might not be for you.
You might consider using [Flint disable comments](/directives) and/or [configuration file disables](/configuration#files) for specific cases instead of completely disabling this rule.

## Further Reading

- [MDN: Classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)
- [TypeScript Handbook: this Types](https://www.typescriptlang.org/docs/handbook/2/classes.html#this-types)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="returnThisTypes" />
