---
description: "Reports usage of substr and substring instead of slice for string operations."
title: "stringSliceMethods"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="stringSliceMethods" />

JavaScript provides three methods for extracting substrings: `slice()`, `substr()`, and `substring()`.
This rule encourages consistent use of `slice()` because:

- `substr()` is deprecated and removed from the ECMAScript specification
- `substring()` has confusing behavior: it auto-swaps arguments when start > end, and treats negative indices as 0
- `slice()` has consistent, predictable behavior that matches `Array.prototype.slice()`

## Examples

### substr to slice

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const result = text.substr(1, 5);
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const result = text.slice(1, 6);
```

</TabItem>
</Tabs>

### substring to slice

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const result = text.substring(1, 5);
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const result = text.slice(1, 5);
```

</TabItem>
</Tabs>

### Negative indices

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const last = text.substring(text.length - 3);
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const last = text.slice(-3);
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you're maintaining legacy code that heavily uses `substr()` or `substring()` and refactoring would be too costly, you may need to disable this rule.
Note that `substr()` is deprecated and may be removed from JavaScript engines in the future.

## Further Reading

- [MDN: String.prototype.slice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice)
- [MDN: String.prototype.substr()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr)
- [MDN: String.prototype.substring()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="stringSliceMethods" />
