---
description: "Reports unnecessary boolean casts."
title: "unnecessaryBooleanCasts"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="unnecessaryBooleanCasts" />

Boolean contexts such as `if` and `while` statements evaluate values for falsiness or truthiness.
Using `!!value` or `Boolean(value)` in those contexts is redundant because that doesn't change whether the value is falsy or truthy.

This rule reports on boolean casts in already-boolean contexts.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
if (!!value) {
	// ...
}
```

```ts
while (Boolean(condition)) {
	// ...
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
if (value) {
	// ...
}
```

```ts
while (condition) {
	// ...
}
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you prefer explicit boolean casts for clarity, this rule may not be for you.

## Further Reading

- [MDN: Boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean)
- [MDN: Logical NOT](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_NOT)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="unnecessaryBooleanCasts" />
